### PR TITLE
feat: support set hardfork from command line

### PR DESCRIPTION
**Pull request by driftluo**
## What this PR does / why we need it?

This PR adds features which support setting hardfork from the command line

like this：
```bash
./target/debug/axon hardfork \
  -c devtools/chain/config.toml \
  --hardfork-start-number 1 --feature f --feature e
```

### What is the impact of this PR?

No Breaking Change

### **PR relation**:
- Ref https://github.com/axonweb3/axon/pull/1404

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [x] E2E Tests
- [x] Web3 Compatible Tests
- [x] OCT 1-5 And 12-15
- [x] OCT 6-10
- [x] OCT 11
- [x] OCT 16-19
- [x] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Chaos CI*                                | Test the liveness and robustness of Axon under terrible network condition |
| *Cargo Clippy*                            | Run `cargo clippy --all --all-targets --all-features`                     |
| *Coverage Test*                           | Get the unit test coverage report                                         |
| *E2E Test*                                | Run end-to-end test to check interfaces                                   |
| *Code Format*                             | Run `cargo +nightly fmt --all -- --check` and `cargo sort -gwc`           |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

**Review by Flouse — COMMENTED** [review]
Is it possible to demonstrate a hardfork workflow in another multi-nodes GitHub action job?
https://github.com/axonweb3/axon/blob/c97f6b06d1c7ee76a0c592f0eea11bfe825098d9/.github/workflows/axon-start-with-short-genesis.yml#L85-L105

File: common/config-parser/src/types/spec.rs [hunk 1]
```diff
@@ -261,24 +261,32 @@ impl Genesis {
             timestamp: self.timestamp,
             // todo: if Hardforkinput is empty, it must change to latest hardfork info to init
```
Comment (Flouse) [inline common/config-parser/src/types/spec.rs:262]: Is it a good time to resolve this TODO?

Comment (driftluo) [reply common/config-parser/src/types/spec.rs:262]: this todo should change every time the hardfork info changed, I feel like there should be a permanent reminder mark here.

**Review by KaoImin — DISMISSED** [review]

File: core/run/src/lib.rs [hunk 2]
```diff
@@ -145,6 +147,15 @@ async fn start<K: KeyProvider>(
     let current_block = storage.get_latest_block(Context::new()).await?;
     let current_state_root = current_block.header.state_root;
 
+    if current_block.header.number
```
Comment (chaoticlonghair) [inline core/run/src/lib.rs]: I think it should throw an error to tell users that your setting is invalid when hardfork number is less than current number.
So, users could change hardfork number a.s.a.p.

File: common/config-parser/src/types/spec.rs [hunk 3]
```diff
@@ -261,24 +261,32 @@ impl Genesis {
...
-pub struct HarforkInput {
-    block_number: u64,
-    hardforks:    Vec<HardforkType>,
+use clap::ValueEnum;
+
+#[derive(Clone, Debug, Deserialize, Args)]
+pub struct HardforkInput {
+    #[arg(long, required = false, requires = "hardforks")]
+    pub block_number: u64,
+    #[arg(long, requires = "block_number")]
```
Comment (KaoImin) [inline common/config-parser/src/types/spec.rs]:
```suggestion
    #[arg(long, requires = "hardfork_start_number")]
```

**Review by chaoticlonghair — COMMENTED** [review]
<details><summary>TL;DR: IMHO, the set operation should be a one-time operation. (click to see details)</summary>

Maybe there should be a new subcommand for this feature.

Looks at the following example:
- Current tip number is 1000.
- The user uses `axon run --block-number 3000 --hardforks feature` to start the chain.
- When the chain reach 2000, the machine has to be restart, so axon is stopped.
- Then, if the user want to restart Axon, he have to use `axon run --block-number 3000 --hardforks feature` to start the chain again.
- After number 3000, the user only have to use `axon run` to start the chain.

So, the user **have to know** the current tip number of the node data (not the tip number of the chain, just the tip number of the node data) before execute `axon run`.
It's not reasonable.

<details><summary>I have a proposal (click to see details).</summary>


- The hardfork features can be set without start the chain.

  `axon hardfork set --number H --feature F`

  if the setting is valid, then save it into persistent caches.
  When the chain is started, read them from the caches.

- We can list the hardfork features which are not reached yet.

  `axon hardfork list`

  Read data from the persistent caches. 

  Maybe we can list all enabled hardfork features, too, if it possible.

  p.s. It could be implemented later, not right now.

- The hardfork features can be removed before the number reached.

  `axon hardfork remove --number H`: remove all hardfork features at `H` if exists.
  `axon hardfork remove --feature F`: remove hardfork feature `F` if exists.

  p.s. It could be implemented later, not right now.

</details>

</details>

@Flouse @KaoImin How about your opinions?

File: core/run/src/lib.rs (hunk 2 above)
Comment (Flouse) [reply core/run/src/lib.rs]: +1

**Comment by KaoImin** [comment]
> 

Noop, the command is an one time operation with the following steps:
1. Stop the chain and download the new binary/image
2. Determine the hardfork start number which must be larger than current tip
3. Start Axon with `--block_number n --hardforks feat_1`

Once the hardfork proposal has become consensus,  no additional arguments need to be added for the following re-start.

**Comment by KaoImin** [comment]
> TL;DR: IMHO, the set operation should be a one-time operation. (click to see details)
> Maybe there should be a new subcommand for this feature.
> 
> Looks at the following example:
> 
> * Current tip number is 1000.
> * The user uses `axon run --block-number 3000 --hardforks feature` to start the chain.
> * When the chain reach 2000, the machine has to be restart, so axon is stopped.
> * Then, if the user want to restart Axon, he have to use `axon run --block-number 3000 --hardforks feature` to start the chain again.
> * After number 3000, the user only have to use `axon run` to start the chain.
> 
> So, the user **have to know** the current tip number of the node data (not the tip number of the chain, just the tip number of the node data) before execute `axon run`. It's not reasonable.
> 
> I have a proposal (click to see details).
> @Flouse @KaoImin How about your opinions?

Noop, the command is an one time operation with the following steps:
1. Stop the chain and download the new binary/image
2. Determine the hardfork start number which must be larger than current tip
3. Start Axon with `--block_number n --hardforks feat_1`

Once the hardfork proposal has become consensus,  no additional arguments need to be added for the following re-start.

**Comment by Flouse** [comment]
> So, the user **have to know** the current tip number of the node data (not the tip number of the chain, just the tip number of the node data) before execute `axon run`. It's not reasonable.

I think it's easy for the chain maintainer to know the current tip number of the node.
As a node maintainer, it's a basic requirement, especially when he/she wants to setup the next hardfork.

---
> - https://github.com/axonweb3/axon/pull/1422#discussion_r1323841843

And I agree with @yangby-cryptape that the `--hardfork_start_number` param should be verified first.

**Comment by chaoticlonghair** [comment]
> Once the hardfork proposal has become consensus, no additional arguments need to be added for the following re-start.

Can you guarantee that the hardfork proposal could become to a consensus feature at once?

Once the chain is stopped, maybe an unintended stop, the users have to decide which command should be used when restart the chain.

> I think it's easy for the chain maintainer to know the current tip number of the node.

Consider that, is any system administrator will restart services manually?

If you write a service to restart axon automatically after it is stopped, you have to check the tip number of the chain data.

- First, you have to modify the service script every time, when a new hardfork feature is on your queue.

- And, how to do fetch a tip number of chain data? Because there is no command to fetch a tip number from chain data without run it at present.
- 
  The only way is to read the tip number from the log files. It doesn't look reliable.

**Comment by KaoImin** [comment]
Some addition points should be added:
1. The number of blocks that each hardfork meet consensus should be recorded.
2. The return of `axon_getHardforkInfo` should be changed as following:
```rust
#[derive(Serialize, Deserialize)]
pub struct HardforksStatus {
    pub enabled:    Vec<HardforkInfo>,
    pub determined: Vec<HardforkInfo>,
    pub proposing:  Vec<HardforkInfo>,
}
```

**Comment by chaoticlonghair** [comment]
> Some addition points should be added:
> 
> ... ...

I prefer to `HashMap<HardforkInfo, Status>` with

```rust
pub enum HardforkStatus {
    Proposed,
    Enabled,
    /.. other statuses ../
}
```

File: core/storage/src/lib.rs [hunk 4]
```diff
@@ -472,6 +474,10 @@ impl<Adapter: StorageAdapter> ReadOnlyStorage for ImplStorage<Adapter> {
             Ok(proof)
         }
     }
+
+    async fn hardfork_proposal(&self, _ctx: Context) -> ProtocolResult<Option<HardforkInfoInner>> {
+        self.adapter.get::<HardforkSchema>(*HARKFORK_PROPOSAL)
```
Comment (KaoImin) [inline core/storage/src/lib.rs]:
```suggestion
        get!(self, *HAEDFORK_PROPOSAL, HardforkSchema)
```

File: core/storage/src/lib.rs [hunk 5]
```diff
@@ -42,6 +43,7 @@ lazy_static::lazy_static! {
     pub static ref LATEST_PROOF_KEY: Hash = Hasher::digest(Bytes::from("latest_proof"));
     pub static ref OVERLORD_WAL_KEY: Hash = Hasher::digest(Bytes::from("overlord_wal"));
     pub static ref MONITOR_CKB_NUMBER_KEY: Hash = Hasher::digest(Bytes::from("monitor_ckb_number"));
+    pub static ref HARKFORK_PROPOSAL: Hash = Hasher::digest(Bytes::from("Hardfork_proposal"));
```
Comment (KaoImin) [inline core/storage/src/lib.rs]:
```suggestion
    pub static ref HARDFORK_PROPOSAL: Hash = Hasher::digest(Bytes::from("hardfork_proposal"));
```

File: common/config-parser/src/types/spec.rs [hunk 6]
```diff
@@ -292,7 +305,7 @@ impl From<HarforkInput> for HardforkInfoInner {
     }
 }
 
-#[derive(Clone, Debug, Deserialize, Copy)]
-enum HardforkType {
+#[derive(Clone, Debug, Serialize, Deserialize, Copy, ValueEnum, EnumIter, PartialEq, Eq, Hash)]
+pub enum HardforkType {
```
Comment (KaoImin) [inline common/config-parser/src/types/spec.rs]: Perhaps named `HardforkName` is more suitable.

```suggestion
pub enum HardforkName {
```

File: core/api/src/jsonrpc/mod.rs [hunk 7]
```diff
@@ -227,7 +227,7 @@ pub trait AxonRpc {
     async fn get_ckb_related_info(&self) -> RpcResult<CkbRelatedInfo>;
 
     #[method(name = "axon_getHardforkInfo")]
-    async fn hardfork_info(&self) -> RpcResult<HardforkInfo>;
+    async fn hardfork_info(&self) -> RpcResult<HashMap<HardforkType, HardforkStatus>>;
```
Comment (KaoImin) [inline core/api/src/jsonrpc/mod.rs]:
```suggestion
    async fn hardfork_infos(&self) -> RpcResult<HashMap<HardforkType, HardforkStatus>>;
```

File: core/consensus/src/engine.rs [hunk 8]
```diff
@@ -98,6 +100,10 @@ impl<Adapter: ConsensusAdapter + 'static> Engine<Proposal> for ConsensusEngine<A
             }
         };
 
+        if remove {
+            self.adapter.remove_hardfork_proposal(ctx.clone()).await?;
```
Comment (KaoImin) [inline core/consensus/src/engine.rs:104]: It should be removed after the execution of the transactions in this block is complete.

File: core/consensus/src/engine.rs [hunk 9]
```diff
@@ -82,13 +82,15 @@ impl<Adapter: ConsensusAdapter + 'static> Engine<Proposal> for ConsensusEngine<A
...
         };
 
+        let mut remove = false;
         let extra_data_hardfork = {
             let mut hardfork = self.node_info.hardfork_proposals.write().unwrap();
             match &*hardfork {
                 Some(v) => {
                     // remove invalid proposal
                     if v.block_number <= next_number {
                         hardfork.take();
```
Comment (KaoImin) [inline core/consensus/src/engine.rs:92]: It should be removed after the execution of the transactions in this block is complete.

File: core/consensus/src/engine.rs (hunk 8 above)
Comment (driftluo) [reply core/consensus/src/engine.rs:104]: This proposal has expired. In this case, it can only be deleted. It is impossible to propose again.

File: core/api/src/jsonrpc/web3_types.rs [hunk 10]
```diff
@@ -875,6 +875,14 @@ pub struct RawLoggerFilter {
     pub topics:     Option<Vec<MultiNestType<Hash>>>,
 }
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub enum HardforkStatus {
+    Proposed,
+    Enabled,
+    Actived,
```
Comment (Flouse) [inline core/api/src/jsonrpc/web3_types.rs]:
```suggestion
    Activated
```

File: core/db/src/rocks.rs [hunk 11]
```diff
@@ -295,6 +296,7 @@ impl From<RocksDBError> for ProtocolError {
 }
 
 // Todo: column family "c0" is reserved for store version
```
Comment (Flouse) [inline core/db/src/rocks.rs]:
```suggestion
// column family "c0" is reserved for store version
```

File: core/run/src/lib.rs [hunk 12]
```diff
@@ -484,3 +492,92 @@ fn execute_transactions(
...
+
+            if info.flags == H256::zero() {
+                return Ok(());
+            }
+
+            let hardforks = metadata_handle.hardfork_infos()?;
+
+            if let Some(v) = hardforks.inner.last() {
+                // if hardfork is already activated, return
+                if v.flags & info.flags == info.flags {
```
Comment (Flouse) [inline core/run/src/lib.rs]:
```suggestion
                if v.flags & info.flags == v.flags {
```
Maybe it's better to rename:
- v -> current_info
- info -> proposed_info

File: core/api/src/jsonrpc/impl/axon.rs [hunk 13]
```diff
@@ -97,12 +99,46 @@ impl<Adapter: APIAdapter + 'static> AxonRpcServer for AxonRpcImpl<Adapter> {
...
+        let current_number = self
+            .adapter
+            .get_block_header_by_number(Default::default(), None)
+            .await
+            .map_err(|e| Error::Custom(e.to_string()))?
+            .unwrap()
+            .number;
+
+        let mut hardfork_infos = HashMap::new();
+        for e in HardforkName::iter() {
```
Comment (Flouse) [inline core/api/src/jsonrpc/impl/axon.rs]: suggestion: rename `e` to `hardfork_name`

File: core/consensus/src/engine.rs (hunk 8 above)
Comment (Flouse) [reply core/consensus/src/engine.rs:104]: Not sure if it is proper to change the state in a `get` function (get_block)?

File: core/run/src/lib.rs (hunk 12 above)
Comment (driftluo) [reply core/run/src/lib.rs]: your opinion is wrong

There are two ways to judge:

`0b11(current) & 0b10(proposal) == 0b10`, feature already activated
`0b11 & 0b01 == 0b01`, feature already activated
`0b11 & 0b100 != 0b100`, feature can proposal

or

`0b11 | 0b10 == 0b11`,feature already activated
`0b11 | 0b01 == 0b11`,feature already activated
`0b11 | 0b100 != 0b11`, feature can proposal

Comment (Flouse) [reply core/run/src/lib.rs]: you are right.

**Review by Flouse — DISMISSED** [review]

File: core/api/src/jsonrpc/impl/axon.rs [hunk 14]
```diff
@@ -97,12 +99,46 @@ impl<Adapter: APIAdapter + 'static> AxonRpcServer for AxonRpcImpl<Adapter> {
...
+                if p.flags & H256::from_low_u64_be((hardfork_name as u64).to_be()) != H256::zero() {
+                    hardfork_infos.insert(hardfork_name, HardforkStatus::Proposed);
+                }
+            }
+            if let Some(a) = actived_latest {
+                if a.flags & H256::from_low_u64_be((hardfork_name as u64).to_be()) != H256::zero() {
+                    if a.block_number <= current_number {
+                        hardfork_infos.insert(hardfork_name, HardforkStatus::Enabled);
+                    } else {
+                        hardfork_infos.insert(hardfork_name, HardforkStatus::Activated);
```
Comment (Flouse) [inline core/api/src/jsonrpc/impl/axon.rs]: suggestion: Proposed -> Determined -> Enabled

File: core/api/src/jsonrpc/impl/axon.rs [hunk 15]
```diff
@@ -97,12 +101,270 @@ impl<Adapter: APIAdapter + 'static> AxonRpcServer for AxonRpcImpl<Adapter> {
...
+            {
+                hardfork_infos.insert(hardfork_name, HardforkStatus::Enabled);
+            }
+        }
+
+        Ok(hardfork_infos)
+    }
+}
+
+fn enabled_and_determined(iter: &[HardforkInfoInner], current_number: u64) -> (H256, H256) {
```
Comment (Flouse) [inline core/api/src/jsonrpc/impl/axon.rs:154]:
```suggestion
/// Returns (enabled_flags, determined_flags) in target block height
fn enabled_and_determined(iter: &[HardforkInfoInner], current_number: u64) -> (H256, H256) {
```

**Review by Flouse — DISMISSED** [review]

**Review by KaoImin — DISMISSED** [review]

File: core/consensus/src/engine.rs [hunk 16]
```diff
@@ -82,13 +82,15 @@ impl<Adapter: ConsensusAdapter + 'static> Engine<Proposal> for ConsensusEngine<A
             RLP_NULL
         };
 
+        let mut remove = false;
         let extra_data_hardfork = {
             let mut hardfork = self.node_info.hardfork_proposals.write().unwrap();
             match &*hardfork {
                 Some(v) => {
                     // remove invalid proposal
```
Comment (Flouse) [inline core/consensus/src/engine.rs]:
```suggestion
                    // remove invalid proposal if the proposed block height is passed
```

**Review by KaoImin — DISMISSED** [review]

**Review by chaoticlonghair — DISMISSED** [review]

**Review by Flouse — DISMISSED** [review]

**Comment by chaoticlonghair** [comment]
I had some review suggestions, but not only about this PR. also for the previous PR of hardfork.

I will create an issue later to discuss them, later.

**Review by KaoImin — APPROVED** [review]

**Review by Flouse — APPROVED** [review]
